api/stripe.go

```diff
@@ -0,0 +1,313 @@
+		"prod_J3m6IuVyh3XOc5": database.TierPremium80,
+	}
+	// stripePricesTest maps Stripe plan prices to specific tiers.
+	// DO NOT USE THESE DIRECTLY! Use stripePrices() instead.
+	stripePricesTest = map[string]int{
+		//"price_1IQAgvIzjULiPWN60U5buItF": database.TierFree,
+		"price_1IReXpIzjULiPWN66PvsxHL4": database.TierPremium5,
+		"price_1IReY5IzjULiPWN6AxPytHEG": database.TierPremium20,
+		"price_1IReYFIzjULiPWN6DqN2DwjN": database.TierPremium80,
+	}
+	// stripePlansProd maps Stripe plans to specific tiers.
+	// DO NOT USE THESE DIRECTLY! Use stripePlans() instead.
+	stripePlansProd = map[string]int{
+		"prod_J2FJE4gMqrOSwn": database.TierFree,
+		"prod_J06NWykm9SRvWw": database.TierPremium5,
+		"prod_J19xHMxmCmBScY": database.TierPremium20,
+		"prod_J19xoBYOMbSlq4": database.TierPremium80,
+	}
+	// stripePricesProd maps Stripe plan prices to specific tiers.
+	// DO NOT USE THESE DIRECTLY! Use stripePrices() instead.
+	stripePricesProd = map[string]int{
+		"price_1IQApHIzjULiPWN6tGNYEIOi": database.TierFree,
+		"price_1IO6AdIzjULiPWN6PtviaWtS": database.TierPremium5,
+		"price_1IP7dMIzjULiPWN6YHoHM3hK": database.TierPremium20,
+		"price_1IP7ddIzjULiPWN6vBhBe9EG": database.TierPremium80,
+	}
+)
+
+type (
+	// stripePrice ...
+	stripePrice struct {
+		ID          string  `json:"id"`
+		Name        string  `json:"name"`
+		Description string  `json:"description"`
+		Tier        int     `json:"tier"`
+		Price       float64 `json:"price"`
+		Currency    string  `json:"currency"`
+		StripeID    string  `json:"stripe"`
+		ProductID   string  `json:"productId"`
+		LiveMode    bool    `json:"livemode"`
+	}
+)
+
+// stripeWebhookHandler handles various events issued by Stripe.
+// See https://stripe.com/docs/api/events/types
+func (api *API) stripeWebhookHandler(w http.ResponseWriter, req *http.Request, _ httprouter.Params) {
+	api.staticLogger.Tracef("Processing request: %+v", req)
+	event, code, err := readStripeEvent(w, req)
+	if err != nil {
+		api.WriteError(w, err, code)
+		return
+	}
+	api.staticLogger.Debugf("Received event: %+v", event)
+
+	// Here we handle the entire class of subscription events.
+	// https://stripe.com/docs/billing/subscriptions/overview#build-your-own-handling-for-recurring-charge-failures
+	// https://stripe.com/docs/api/subscriptions/object
+	if strings.Contains(event.Type, "customer.subscription") {
+		var s stripe.Subscription
+		err = json.Unmarshal(event.Data.Raw, &s)
+		if err != nil {
+			api.staticLogger.Warningln("Failed to parse event. Error: ", err, "\nEvent: ", string(event.Data.Raw))
+			return
+		}
+		err = api.processStripeSub(req.Context(), &s)
+		if err != nil {
+			api.staticLogger.Debugln("Failed to process sub:", err)
+		}
+		api.WriteSuccess(w)
+		return
+	}
+
+	// Here we handle the entire class of subscription_schedule events.
+	// See https://stripe.com/docs/api/subscription_schedules/object
+	if strings.Contains(event.Type, "subscription_schedule") {
```

Comment:
q: Should this be `custom.subscription_schedule` to match `custom.subscription` above?

Reply:
Those are two different groups of events - `customer` and `subscription_schedule` (https://stripe.com/docs/api/events/types). Both can be sent on subscription update, so I'm listening to both.